grr/server/grr_response_server/gui/api_plugins/metadata.py

```diff
@@ -716,58 +887,68 @@ def _GetArraySchema(items_type_name: str) -> ArraySchema:
+  that represent a map entry (i.e. auxiliary protobuf messages with a `key` and
+  `value` field). This function verifies that all the signs of a compiled
+  `protobuf.map` field are present for the current `FieldDescriptor` and, if
+  this field is actually a `protobuf.map`, it returns the `key` and `value`
+  `FieldDescriptor`s.
 
   Args:
-    field_descriptor: The protobuf `FieldDescriptor` associated with the target
-      field.
+    field_descriptor: The protobuf `FieldDescriptor` whose type is checked if it
+      is a map and whose type's associated `key` and `value` `FieldDescriptor`s
+      are returned.
 
   Returns:
-    If the field is not part of a `protobuf.oneof`, then a simple `Reference
-    Object`, else, a dictionary that includes a description entry along the
-    `Reference Object`.
+    A `KeyValueDescriptor` named tuple consisting of the `key` and `value`
+    `FieldDescriptor`s (in this order) extracted from the given
+    `FieldDescriptor`'s map entry message type, or `None` if the given
+    `FieldDescriptor` does not describe a map field.
   """
-  # The semantic of `protobuf.oneof` is not currently supported by the
-  # OpenAPI Specification. See this GitHub issue [1] for more details.
-  #
-  # [1]: github.com/google/grr/issues/822
-  type_name = _GetTypeName(field_descriptor)
-  containing_oneof: OneofDescriptor = field_descriptor.containing_oneof
-  if containing_oneof is None:
-    if field_descriptor.label == protobuf2.LABEL_REPEATED:
-      return _GetArraySchema(type_name)
-
-    return _GetReferenceObject(type_name)
-
-  # The following `allOf` is required to display the description by
-  # documentation generation tools because the OAS specifies that there
-  # should not be any sibling properties to a `$ref`. This is the
-  # workaround proposed by the ReDoc community [1].
-  #
-  # [1]: github.com/Redocly/redoc/issues/453#issuecomment-420898421
-  oneof_member = cast(
-    OneofFieldSchema,
-    {
-      "description":
-        f"This field is part of the \"{containing_oneof.name}\" oneof. "
-        f"Only one field per oneof should be present.",
-    }
-  )
-  if field_descriptor.label == protobuf2.LABEL_REPEATED:
-    oneof_member["allOf"] = [
-      _GetArraySchema(type_name),
-    ]
-  else:
-    oneof_member["allOf"] = [
-      _GetReferenceObject(type_name),
-    ]
+  if field_descriptor.label != protobuf2.LABEL_REPEATED:
+    return None
+
+  entry_descriptor: Optional[Descriptor] = field_descriptor.message_type
+  if entry_descriptor is None:
+    return None
+
+  if _GetMapEntryTypeName(field_descriptor.name) != entry_descriptor.name:
+    return None
+
+  if len(entry_descriptor.fields) != 2:
+    return None
+
+  if (
+      entry_descriptor.fields[0].name == "key" and
+      entry_descriptor.fields[1].name == "value"
+  ):
+    return KeyValueDescriptor(
+      entry_descriptor.fields[0], entry_descriptor.fields[1]
+    )
```

Comment:
Prefer to use keyword arguments when initializing struct-like things. This makes the code more resilient to changes like re-ordering the fields. Here we use `NamedTuple` because data classes were introduced in Python 3.7, so we cannot use them yet.

Reply:
Thank you for this suggestion! I made the change.